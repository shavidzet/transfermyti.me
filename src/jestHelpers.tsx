/* eslint-disable import/no-extraneous-dependencies,@typescript-eslint/no-explicit-any,no-console */
import React from "react";
import {
  RenderOptions,
  RenderResult,
  render as testRender,
} from "@testing-library/react";
import { ThemeProvider } from "emotion-theming";
import { theme } from "./components/emotion";

// These are exported for convenience (so we don't have to add an additional import
// to the test file).
export { createEvent, fireEvent, RenderResult } from "@testing-library/react";
export { theme } from "./components/emotion";

/**
 * Renders the specified component with react-testing-library wrapped in a theme
 * provider so we don't have to do it explicitly for each component.
 */
export function render(
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "queries">,
): RenderResult {
  return testRender(<ThemeProvider theme={theme}>{ui}</ThemeProvider>, options);
}

type LogMethod = "log" | "warn" | "error";

export class ConsoleToggle {
  private consoleMethod: (...args: any[]) => void = console.warn;
  private readonly logMethod: LogMethod;

  public constructor(logMethod: LogMethod) {
    this.logMethod = logMethod;
  }

  public onBefore(): void {
    this.consoleMethod = console[this.logMethod];
    console[this.logMethod] = jest.fn();
  }

  public onAfter(): void {
    console[this.logMethod] = this.consoleMethod;
  }
}
