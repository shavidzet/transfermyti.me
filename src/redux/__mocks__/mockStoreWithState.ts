import { ActionCreator, Middleware } from "redux";
import configureStore, { MockStore } from "redux-mock-store";
import createSagaMiddleware from "redux-saga";
import { default as stateFixture } from "../__fixtures__/state";

export { MockStoreCreator, MockStore } from "redux-mock-store";

// This is the same as ReduxAction, but the type was changed to `any`, otherwise
// we'd have to explicitly specify every action `type` name when using jest-in-case:
export interface ReduxTestAction<TPayload = {}> {
  type: any;
  payload?: TPayload;
  error?: boolean;
  meta?: any;
}

export type ReduxTestActionCreator<TPayload> = ActionCreator<
  ReduxTestAction<TPayload>
  >;

// This is done to simplify the imports in tests (rather than have to import
// state/fixtures from the state.ts file and mockStoreWithState() function from
// another:
export const state = stateFixture;

/**
 * Creates an instance of a Redux store for mocking in tests.
 * @param stateChanges Updated entries in state (doesn't have to be entire state
 *                     since state fixture is used.
 * @param additionalMiddleware Any other middlewares to pass to configureStore
 *                             function.
 */
export function mockStoreWithState<TState>(
  stateChanges?: TState,
  additionalMiddleware: Array<Middleware> = [],
): MockStore {
  const sagaMiddleware = createSagaMiddleware();
  const mockStore = configureStore([sagaMiddleware, ...additionalMiddleware]);
  const validState = { ...stateFixture, ...stateChanges };
  return mockStore(validState);
}
