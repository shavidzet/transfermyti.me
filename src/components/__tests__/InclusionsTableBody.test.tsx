import React from "react";
import { render, RenderResult } from "~/jestHelpers";
import { state } from "~/redux/__mocks__/mockStoreWithState";
import { clientsForInclusionsTableSelector } from "~/clients/clientsSelectors";
import InclusionsTableBody from "../InclusionsTableBody";

const setup = (
  propOverrides: any = {},
): { props: any; wrapper: RenderResult } => {
  const props = {
    fieldNames: ["name", "entryCount", "projectCount"],
    tableData: clientsForInclusionsTableSelector(state),
    onFlipIsIncluded: jest.fn(),
    ...propOverrides,
  };

  const table = document.createElement("table");

  const wrapper = render(<InclusionsTableBody {...props} />, {
    container: document.body.appendChild(table),
  });

  return { props, wrapper };
};

describe("the <InclusionsTableBody> component", () => {
  test("matches its snapshot with valid props", () => {
    const { wrapper } = setup();

    expect(wrapper.container.firstElementChild).toMatchSnapshot();
  });
});
