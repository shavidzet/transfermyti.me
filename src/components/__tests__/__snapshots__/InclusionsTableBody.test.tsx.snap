// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the <InclusionsTableBody> component matches its snapshot with valid props 1`] = `
.emotion-0 td {
  color: rgb(136,136,136);
  -webkit-text-decoration: line-through;
  text-decoration: line-through;
}

<tbody>
  <tr
    class="emotion-0"
  >
    <td>
      Test Client A
    </td>
    <td>
      3
    </td>
    <td>
      1
    </td>
    <td
      data-include="true"
    >
      <input
        disabled=""
        type="checkbox"
      />
    </td>
  </tr>
  <tr
    class="emotion-0"
  >
    <td>
      Test Client B
    </td>
    <td>
      3
    </td>
    <td>
      1
    </td>
    <td
      data-include="true"
    >
      <input
        disabled=""
        type="checkbox"
      />
    </td>
  </tr>
</tbody>
`;
